test/integration/client/connection_test.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"os/exec"
+	"regexp"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	utildbus "k8s.io/kubernetes/pkg/util/dbus"
+	"k8s.io/kubernetes/pkg/util/iptables"
+	"k8s.io/kubernetes/test/integration/framework"
+	utilsexec "k8s.io/utils/exec"
+)
+
+func TestReconnectBrokenTCP(t *testing.T) {
+	// TODO: remove env setting when kubernetes/client-go#374 is fixed
+	// use http to detect non-responding tcp connection
+	defer func() {
+		if err := os.Setenv("DISABLE_HTTP2", ""); err != nil {
+			t.Fatalf("failed to re-enable http2: %v", err)
+		}
+	}()
+	if err := os.Setenv("DISABLE_HTTP2", "true"); err != nil {
+		t.Fatalf("failed to disable http2: %v", err)
+	}
+
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, nil, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	client := clientset.NewForConfigOrDie(server.ClientConfig).CoreV1().Endpoints("default")
+	w, err := client.Watch(metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		t.Fatalf("failed to watch endpoints: %v", err)
+	}
+
+	// find port that talks to test server
+	connections, err := exec.Command("ss", "-tn").Output()
+	if err != nil {
+		t.Fatalf("failed ot list tcp connections: %v", err)
+	}
+
+	u, err := url.Parse(server.ClientConfig.Host)
+	if err != nil {
+		t.Fatalf("failed to parse url %s: %v", server.ClientConfig.Host, err)
+	}
+	pattern := regexp.MustCompile(fmt.Sprintf(`%s[\t ]+[0-9.]+:([0-9]+)`, u.Host))
+	ports := pattern.FindStringSubmatch(string(connections))
+	sport := ports[1]
```

Comment:
@caesarxuchao pointed out that in HTTP/1.1 scenario there are multiple ports talking to apiserver and this line doesn't necessarily capture the watch connection